### ISSUE TITLE
perf(agent): coalesce valuable cold CPU shared prefixes in the native planner

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12749]
# perf(agent): coalesce concurrent cold CPU shared-prefix prefill

Parent issue: https://github.com/anthony-chaudhary/fak/issues/1914

Implementation issue: https://github.com/anthony-chaudhary/fak/issues/12749

Status: implemented and independently tested in public commit `2302e65ebae6de36f74411836ca43bce38665b18`, based on `eba3edad2563fea3707dc389cc8120ce42bba901`. The source finding is tracked in [the parent issue update](https://github.com/anthony-chaudhary/fak/issues/1914#issuecomment-5624030306); #12749 is its bounded implementation child. Final focused race and unfiltered affected-package checks passed. A CPU experiment rejected unconditional multicore admission; the final path is constrained to one Go execution core. Canonical landed blobs match all five reviewed files and context provenance is `SEPARATED`. The independent original/repaired active-planner witness and validation receipts are retained with the project; the [public regression](https://github.com/anthony-chaudhary/fak/blob/2302e65ebae6de36f74411836ca43bce38665b18/internal/agent/inkernel_cpu_prefix_flight_test.go) and commands below are reproducible.

```routing
lane: agent
paths: ["internal/radixkv/singleflight.go", "internal/radixkv/singleflight_test.go", "internal/agent/inkernel_planner.go", "internal/agent/inkernel_decode.go", "internal/agent/inkernel_cpu_prefix_flight_test.go"]
expected_steps: 6
```

## Current state

On baseline `eba3edad2563fea3707dc389cc8120ce42bba901`, the ordinary `InKernelPlanner` CPU path performs an authorized radix lookup, releases the cache mutex, prefills a request-local session, and admits the completed KV afterward. Concurrent sibling requests can therefore all observe a cold miss before any sibling admits its prompt. Each then evaluates the same long prefix independently.

`radixkv.PrefixFlightGroup` is implemented and tested, but production does not retain or call one. Its standalone `CoalesceNS` lookup is keyed by namespace plus the complete requested token sequence. It coalesces identical prompts, but distinct `P+suffixA`, `P+suffixB`, and `P+suffixC` keys cannot join even when the requests overlap while `P` is being evaluated. `ScopedTree.FlightGroup` also creates a new group on every call, which cannot coordinate independent planner requests unless a caller retains the returned group.

Source witness: `internal/agent/inkernel_decode.go:103-218` completes cache lookup and releases `p.mu`; `internal/agent/inkernel_decode.go:269-303` performs prefill; `internal/agent/inkernel_decode.go:323-352` admits the completed CPU KV. `internal/radixkv/singleflight.go:108-121` encodes an exact token key, and `internal/radixkv/singleflight.go:133-206` only joins that exact key. No production caller of `CoalesceNS`, `CoalescePrefix`, or `ForkSuffix` exists outside radixkv.

## Problem frame

Centrality: Core

P1: advanced - concurrent local CPU agent requests with a valuable shared prompt prefix execute that prefix once and retain independent suffix/decode state.

P2: advanced - callback counts and matched-token accounting prove removed duplicate CPU prefill work; repeated CPU cohort timings and clone-byte accounting decide admission. No accelerator throughput claim follows.

P3: preserved - authenticated tenant isolation, request cancellation, sampling state, output order, radix admission, and device serialization remain unchanged.

P4: preserved - multicore, cache-disabled, device, native Metal, speculative target/draft, recurrent caches that cannot truncate, and short-prefix behavior retain their current paths.

## Parent context

#1914 proposes broad cross-agent prefill fusion and identical-request output fanout. This ticket is its bounded CPU implementation slice: transient KV-prefix singleflight for distinct sibling prompts in the active planner, without fused kernels, decode batching, or output sharing. Closed #11633 supplied the radix singleflight primitive but did not wire the planner and its exact-token key does not join divergent suffixes. Open #12618 and #12629 cover tree-backed result ownership defects; this slice uses a standalone planner-owned group and must still return independent KV clones.

## Why now

The local runtime already has persistent radix reuse and an unused in-flight coordination primitive. Closing the cold overlap between lookup and admission is a small active-path change with a deterministic software oracle. This ticket makes a CPU-only claim and supplies no Strix or GPU performance evidence.

## Core through-line

After normal authorized lookup has settled on a cold plain-CPU miss, register the request's namespace and complete prompt in one planner-owned transient flight group only when `runtime.GOMAXPROCS(0) == 1` and the initialized session cache supports safe tail eviction. A later request in the same namespace may wait for an active prompt when its common prefix is at least 64 tokens and the leader's divergent suffix is no longer than that prefix. Among eligible leaders, prefer the greatest common-prefix-minus-leader-suffix token surplus, then the longer common prefix, then the earlier registration. This bounds wasted waiting by token work; it is not a general latency model.

When the leader completes prefill, publish an immutable KV copy before leader decode or follower wakeup. Each follower receives its own clone, safely evicts the leader-only tail including auxiliary cache state and lineage, clears partial-match logits, and evaluates its own divergent suffix. A shorter exact follower refeeds its own last token to produce the correct logits. Each request then admits its own full prompt and performs its own sampling and decode. Track numeric KV payload bytes copied; skip publication copying when no live followers remain.

The 64-token floor uses the existing radix/checkpoint block scale. It is a constant, not a new flag. Dynamic thresholds, worker-pressure admission, and multicore opt-in configuration remain outside this leaf.

Admission limitation: token savings alone do not establish a latency benefit. A repeated tiny-model CPU experiment found useful one-core cohort savings but a multicore cohort regression, so automatic admission is restricted to the measured one-core regime. The independent witness also includes a leader with a long divergent suffix, a lone cold request, and host-default bypass. Final receipts must retain both the rejected unconditional candidate and the final policy. Larger models and other execution regimes remain unqualified.

## Working spine

Authenticated local request -> ordinary authorized radix miss -> same-tenant CPU prefix-flight election -> one leader prefill -> immutable KV handoff -> independent follower KV clone truncated to the longest common prefix -> each suffix prefilled separately -> existing full-prompt radix admission -> independent decode and output.

## Gold-plating boundary

No device or native Metal integration. No speculative target/draft integration. No decode batching, output fanout, sampling deduplication, scheduler rewrite, new public configuration, persistent cache-scope promotion, GPU claim, or arbitrary short-prefix merging. Do not hold `p.mu` or the flight-group mutex during prefill, waiting, KV cloning, suffix computation, admission, or decode. Do not change the existing device execution lock or #12743 Vulkan decode batching work.

## Concrete repro witness

Add a counting/blocking CPU session fixture and three distinct prompts `P+suffixA`, `P+suffixB`, and `P+suffixC`, where `len(P) >= 128` and every suffix is longer than 64 tokens. The leader callback closes an `entered` channel and blocks on `release`. Start B and C only after `entered`; wait with a bounded logical predicate until both have joined the flight, then close `release`. Do not use sleep ordering as the oracle.

Executed active-planner negative control: the same source-neutral Go test overlay and `GOMAXPROCS=1` ran against original `eba3edad2` and the repaired source. The existing cache mutex aligns request arrival; no production prefill/cache logic is replaced. The original fails with `matched=[0 0 0]`; the repair passes with `matched=[0 512 512]`. Prompt-minus-returned-match accounting changes from 1,824 to 800 tokens. This is derived accounting, not an instrumented operation counter. Mutex arrival alignment is an observed concurrency reproduction, not a deterministic barrier after every cold lookup.

Separate deterministic helper and planner tests hold a registered leader, wait until both followers have joined, then release it. They verify callback coalescing, actual matched depths, per-follower KV ownership, independent suffix output parity, exact shorter-prompt refeed, cancellation cleanup, namespace isolation, admission ranking, and bypass conditions. The complete active path is also measured through `InKernelPlanner.Complete` with cold output oracles.

```sh
GOWORK=off go test -race ./internal/agent ./internal/radixkv -run '^(TestInKernelCPUConcurrentPrefixFlight|TestPrefixFlightGroupSharedPrefixContract)$' -count=1 -timeout 5m
```

## Verifiable Witness

The tests use channels and callback counters rather than elapsed time:

1. The leader-registration barrier proves B and C arrive while A is in flight. A bounded join predicate proves both followers registered before release.
2. The full-prompt callback count is one for the shared-prefix cohort. B and C report `matched == len(P)`.
3. Distinct suffix operation counts prove suffix A, B, and C are each evaluated. Partial flight hits expose no cached logits before suffix prefill.
4. Greedy generated token IDs for A, B, and C equal cache-disabled full recomputation.
5. Cancelling B while it waits returns `context.Canceled`; A and C finish, the flight drains, and no leader work is cancelled by B.
6. A live context retries election at most once if its joined leader fails or is cancelled; no stale promise or deadlock remains.
7. The same tokens under another authenticated tenant execute a separate leader callback and observe no cross-tenant match.
8. Prompts whose longest common prefix is shorter than 64 tokens do not join and retain one callback per request.

## Exact file:line seams

- `internal/agent/inkernel_planner.go:78-112` (`InKernelPlanner` cache and device synchronization fields): retain one lazily initialized standalone `PrefixFlightGroup` for the planner lifetime; use `sync.Once` or constructor initialization without changing public APIs.
- `internal/agent/inkernel_decode.go:103-218` (`generateReusedContextWithBias` lookup): derive an opaque transient namespace from the existing authenticated tenant identity, with a separate fixed legacy namespace for unscoped calls.
- `internal/agent/inkernel_decode.go:269-352` (`generateReusedContextWithBias` prefill/admission): elect or join only when reuse is enabled, the settled match is cold, and `!p.requiresDeviceSerialization()`. A leader publishes an immutable KV clone; a follower restores only its longest common prefix and always clears logits for a partial match before normal suffix prefill and admission.
- `internal/agent/inkernel_planner.go:1864-1874,2176-2178` (`Complete` / `requiresDeviceSerialization`): preserve the device boundary; the new flight is unreachable for backend and native-Metal requests.
- `internal/radixkv/singleflight.go:64-76,108-206` (`PrefixFlightGroup`, `prefixFlightKey`, `CoalesceNS`): add longest-active-prefix selection within an exact namespace, independent clone ownership, a 64-token minimum value floor, cancellable follower waiting, prompt metadata cleanup, and bounded follower re-election after leader failure.

## Blast radius and fallback

Only public `internal/agent` and `internal/radixkv` are affected. The new mechanism is gated by existing CPU prefix-reuse eligibility and a cold lookup result. Cache-disabled requests, device/native-Metal paths, short common prefixes, and unsupported models follow the current full-prefill behavior. A flight error wakes every follower; a still-live follower may retry once and otherwise returns the existing error. Persistent radix admission remains the durable fallback after the transient flight ends. No unrelated test is skipped or quarantined.

## Scoped Acceptance Criteria

- [x] Three barrier-aligned distinct CPU prompts with a shared prefix of at least 128 tokens execute one flight callback and independently execute all three suffixes.
- [x] Each request's greedy token IDs equal cache-disabled full recomputation; sampling RNG, stops, penalties, streaming callbacks, and decode are per request.
- [x] Every follower owns a KV clone truncated to its actual common-prefix length; partial matches return no cached logits.
- [x] Tenant B cannot join tenant A's flight. Unscoped legacy calls cannot join a scoped tenant flight.
- [x] Cancelling one follower does not cancel the leader or another follower, leak an in-flight entry, or strand a waiter.
- [x] Leader cancellation, error, and panic close the promise; a live follower retries election at most once.
- [x] Common prefixes shorter than 64 tokens and leaders whose divergent suffix exceeds the shared prefix do not coalesce.
- [x] All followers cancelling before leader completion leaves no publication-copy cost or stale waiter.
- [x] Multicore, cache-disabled, device, native-Metal, speculative target/draft, and non-evictable recurrent-cache paths retain their existing behavior.
- [x] Repeated one-core and host-default CPU cohorts preserve output parity, report P50/P90 and uncertainty, and account separately for derived computed tokens, actual restored tokens, and numeric KV clone bytes. The final arrival-corrected receipt uses the same minimum delay in both arms and retains every sample.
- [x] The focused race witness and full `internal/agent` plus `internal/radixkv` package tests pass on exact landed `2302e65e`. Final unfiltered gate: agent 133.665s, radixkv 1.775s; no exclusions. Final focused race: agent 4.542s, radixkv 13.599s. Existing unrelated timing tests were not edited.

## Done condition

The independent active-planner regression fails on the original source with duplicate work and passes with the repair. The radix unit test proves value-ranked active-prefix selection, clone isolation, namespace isolation, cancellation cleanup, error re-election, and the admission floor. Repeated CPU timing qualifies only the implemented one-core default; multicore bypass avoids the observed waiting regression. A verified public commit references #1914 or a linked child issue and carries the exact test receipt. No accelerator performance claim is made from this software witness.

## Witness

```sh
GOWORK=off go test -race ./internal/agent ./internal/radixkv -run '^(TestInKernelCPUConcurrentPrefixFlight|TestPrefixFlightGroupSharedPrefixContract)$' -count=1 -timeout 5m
GOWORK=off go test ./internal/agent ./internal/radixkv -count=1 -timeout 5m
```

## Acceptance gate

The focused race command is green; the affected packages are green; the public boundary/leak audits are green; and the before-fix receipt records duplicate CPU prefill callback counts on the original SHA. Hardware qualification is not part of this CPU software leaf.

## Closure binding

One verified public commit cites #1914 or its explicitly linked child issue, carries `(fak agent)`, records independent implementation and test contexts, and preserves the original-SHA negative-control plus repaired-SHA race receipts.

## Likely files

- `internal/radixkv/singleflight.go`
- `internal/radixkv/singleflight_test.go`
- `internal/agent/inkernel_planner.go`
- `internal/agent/inkernel_decode.go`
- `internal/agent/inkernel_cpu_prefix_flight_test.go`

## Lane

agent; narrow supporting transient coordination in radixkv. Parent: #1914. Related completed primitive: #11633. Ownership follow-ups: #12618 and #12629. Coordinate with #12743; this ticket does not alter Vulkan decode batching.

## Expected steps

1. Add an independent radix regression for longest active same-namespace prefix selection, the 64-token floor, clone ownership, cancellation, and error cleanup.
2. Extend standalone `PrefixFlightGroup` with longest-active-prefix flight metadata and bounded re-election while preserving exact-key APIs.
3. Retain one group on `InKernelPlanner` and derive tenant-safe transient namespaces from the existing authenticated context.
4. Wire the ordinary cold CPU lookup-to-prefill seam; keep device, Metal, cache-disabled, and speculative routes unchanged.
5. Add the barrier-controlled planner integration regression with callback, suffix, parity, cancellation, and tenant assertions.
6. Run focused race, affected-package, boundary, context-separation, and leak checks; record the original and repaired SHA receipts.

**Comment by anthony-chaudhary** [comment]
Implemented in [2302e65ebae6de36f74411836ca43bce38665b18](https://github.com/anthony-chaudhary/fak/commit/2302e65ebae6de36f74411836ca43bce38665b18), verified present on public main.

The normal planner now coalesces valuable concurrent cold CPU prefixes within the existing tenant boundary. Followers receive independent KV state, safely discard the leader-only tail, recompute their suffix, and retain separate decode. Cancellation drains active waiters without cancelling peers or cloning for an abandoned cohort.

Automatic admission is limited to GOMAXPROCS=1, supported evictable CPU caches, and a common prefix of at least 64 tokens whose saved work covers the leader-only suffix. Multicore, device, recurrent and speculative paths retain existing behavior. A controlled CPU experiment rejected unconditional multicore admission; no general latency or hardware claim is made.

Independent original active-planner control reports matched=[0,0,0]; the repaired same-overlay run reports [0,512,512]. Exact landed-source gates passed with HEAD unchanged and no exclusions:

```sh
GOWORK=off go test -race ./internal/agent ./internal/radixkv -run '^(TestInKernelCPUConcurrentPrefixFlight|TestPrefixFlightGroupSharedPrefixContract)$' -count=1 -timeout 5m
GOWORK=off go test ./internal/agent ./internal/radixkv -count=1 -timeout 10m
```

Five reviewed file blobs match the landed source; boundary/leak checks pass and implementation/test context provenance is SEPARATED. This closes the bounded CPU child. Parent #1914 remains open for broader fusion and request-coalescing work.
